test/inductor/test_distributed_patterns.py

```diff
@@ -395,6 +395,11 @@ def fn(x):
         self._assert_same_grad(r1, r2)
         self._assert_same_grad(p1, p2)
 
+    # TODO: fix fw_metadata.num_forward in dynamo.
+    # While fn returns two user outputs, dynamo only marks one user outputs
+    # p.sin(). This leads to p being wrongly marked as donated buffer
+    # and inplace updated.
+    @torch._functorch.config.patch(donated_buffer=False)
     def test_nn_param_return3(self):
```

Comment:
if it doesn't support it, does it graph break ?  this is valid user code do we give an incorrect result ?

Reply:
I dont fully understand the interaction with this PR. But there is something weird about creating `nn.Parameter` in the compiled frame. There was a concept of sacrificial parameter by @jansel and @yf225.

In this case, we do have a single graph from Dynamo. Just that Dynamo Fx graph does not return the parameter here, it just returns p.sin(). In the modified residual bytecode, Dynamo returns both parameter and p.sin(). So, its all correct. 

https://www.internalfb.com/phabricator/paste/view/P1682708531

Reply:
Looks like the donated buffer analysis is not correctly accounting for `torch.ops.aten.set_.source_Tensor(primals_2, add); `. can we fix this before landing @BoyuanFeng

Reply:
This would also be fixed if we made aot_autograd return the result of `set_` and drop it, which is something I think we should do anyway. we had another issue in cudagraphs that would have also been fixed by this... cc @bdhirsh

Reply:
My recollection is that when we first added the sacrificial parameter logic, we **needed** the set_ to be in the graph in order to let inductor free memory at runtime (and not keep it live til the end of the region).

Maybe a question for @yf225: can/should we just kill this sacrificial parameter code entirely? Or do you think we still need it (for FSDP2 tracing).

I guess one stop-gap alternative if we don't think we can kill it is just to shut off this config whenever dynamo detects a sacrificial parameter in the graph, although that doesn't feel like a great long-term state

Reply:
I think we won't use the sacrificial parameter code for FSDP2 tracing (instead we run 1st iteration with eager to init all parameters in eager, and then in 2nd iteration we start doing compile).

If it's tricky to maintain sacrificial parameter code, I'd vote for graph-breaking on nn.Parameter creation instead.